<a href="https://colab.research.google.com/github/navinsambari123/Mobile-price-range-prediction/blob/main/Github_Mobile_price_range_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PROBLEM STATEMENT**

##In the competitive mobile phone market companies want to understand sales data of mobile phones and factors which drive the prices. The objective is to find out some relation between features of a mobile phone(eg:- RAM, Internal Memory, etc) and its selling price. In this problem, we do not have to predict the actual price but a price range indicating how high the price is.




---



# **DATA DESCRIPTION**

##1. Battery_power - Total energy a battery can store in one time measured in mAh
##2. Blue - Has bluetooth or not
##3. Clock_speed - speed at which microprocessor executes instructions
##4. Dual_sim - Has dual sim support or not
##5. Fc - Front Camera mega pixels
##6. Four_g - Has 4G or not
##7. Int_memory - Internal Memory in Gigabytes
##8. M_dep - Mobile Depth in cm
##9. Mobile_wt - Weight of mobile phone
##10. N_cores - Number of cores of processor
##11. Pc - Primary Camera mega pixels
##12. Px_height - Pixel Resolution Height
##13. Px_width - Pixel Resolution Width
##14. Ram - Random Access Memory in Mega Bytes
##15. Sc_h - Screen Height of mobile in cm
##16. Sc_w - Screen Width of mobile in cm
##17. Talk_time - longest time that a single battery charge will last when you are
##18. Three_g - Has 3G or not
##19. Touch_screen - Has touch screen or not
##20. Wifi - Has wifi or not
##21. Price_range - This is the target variable with value of 0(low cost), 1(medium cost), 2(high cost) and 3(very high cost).



---



#**IMPORTING THE LIBRARIES**

In [1]:
# IMPORTING THE LIBRARIES AND MODULES
import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.model_selection import cross_val_score 
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV
from sklearn.svm import SVC 
from sklearn.pipeline import make_pipeline

from sklearn.tree import DecisionTreeClassifier

from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import AdaBoostClassifier
from xgboost import XGBClassifier

# IMPORTING METRICS RELATED MODULE FROM SKLEARN LIBRARY
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score,plot_confusion_matrix,classification_report,roc_auc_score,roc_curve, auc,confusion_matrix
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
# MOUNTING WITH GOOGLE DRIVE 
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# LOADING THE DATASET
data = pd.read_csv("/content/drive/MyDrive/Mobile Price Range Prediction - Navinkumar Sambari/mobile_price_range.csv")

In [4]:
# LET SEE THE FIRST FIVE ROWS OF THE DATASET
data.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

In [5]:
# LAST FIVE ROWS OF THE DATASET
data.tail()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
1995            794     1          0.5         1   0       1           2   
1996           1965     1          2.6         1   0       0          39   
1997           1911     0          0.9         1   1       1          36   
1998           1512     0          0.9         0   4       1          46   
1999            510     1          2.0         1   5       1          45   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
1995    0.8        106        6  ...       1222      1890   668    13     4   
1996    0.2        187        4  ...        915      1965  2032    11    10   
1997    0.7        108        8  ...        868      1632  3057     9     1   
1998    0.1        145        5  ...        336       670   869    18    10   
1999    0.9        168        6  ...        483       754  3919    19     4   

      talk_time  three_g  touch_screen  wifi  price_range  
1995         19        1             1     0            0  
1996         16        1             1     1            2  
1997          5        1             1     0            3  
1998         19        1             1     1            0  
1999          2        1             1     1            3  

[5 rows x 21 columns]

In [6]:
# DATASET INFORMATION
def information(data):
    print(f"Dataset Shape: {data.shape}")
    information = pd.DataFrame(data.dtypes,columns=['dtypes']).reset_index()
    information['Name'] = information['index']
    information= information[['Name','dtypes']]
    information['Missing'] = data.isnull().sum().values    
    information['Uniques'] = data.nunique().values
    return information
result = information(data).sort_values('Missing', ascending= False)
print(result)

Dataset Shape: (2000, 21)
             Name   dtypes  Missing  Uniques
0   battery_power    int64        0     1094
11      px_height    int64        0     1137
19           wifi    int64        0        2
18   touch_screen    int64        0        2
17        three_g    int64        0        2
16      talk_time    int64        0       19
15           sc_w    int64        0       19
14           sc_h    int64        0       15
13            ram    int64        0     1562
12       px_width    int64        0     1109
10             pc    int64        0       21
1            blue    int64        0        2
9         n_cores    int64        0        8
8       mobile_wt    int64        0      121
7           m_dep  float64        0       10
6      int_memory    int64        0       63
5          four_g    int64        0        2
4              fc    int64        0       20
3        dual_sim    int64        0        2
2     clock_speed  float64        0       26
20    price_range    int64   

In [7]:
# CHECKING DUPLICATE RECORDS
print(f'{data.duplicated().sum()} duplicate values in our dataset.')

0 duplicate values in our dataset.


In [8]:
# UNIQUE VALUES OF EACH COLUMN
for i in data.columns:
  print(f"UNIQUE VALUES of {i}")
  print(data[i].unique())
  print("\n")

UNIQUE VALUES of battery_power
[ 842 1021  563 ... 1139 1467  858]


UNIQUE VALUES of blue
[0 1]


UNIQUE VALUES of clock_speed
[2.2 0.5 2.5 1.2 1.7 0.6 2.9 2.8 2.1 1.  0.9 1.1 2.6 1.4 1.6 2.7 1.3 2.3
 2.  1.8 3.  1.5 1.9 2.4 0.8 0.7]


UNIQUE VALUES of dual_sim
[0 1]


UNIQUE VALUES of fc
[ 1  0  2 13  3  4  5  7 11 12 16  6 15  8  9 10 18 17 14 19]


UNIQUE VALUES of four_g
[0 1]


UNIQUE VALUES of int_memory
[ 7 53 41 10 44 22 24  9 33 17 52 46 13 23 49 19 39 47 38  8 57 51 21  5
 60 61  6 11 50 34 20 27 42 40 64 14 63 43 16 48 12 55 36 30 45 29 58 25
  3 54 15 37 31 32  4 18  2 56 26 35 59 28 62]


UNIQUE VALUES of m_dep
[0.6 0.7 0.9 0.8 0.1 0.5 1.  0.3 0.4 0.2]


UNIQUE VALUES of mobile_wt
[188 136 145 131 141 164 139 187 174  93 182 177 159 198 185 196 121 101
  81 156 199 114 111 132 143  96 200  88 150 107 100 157 160 119  87 152
 166 110 118 162 127 109 102 104 148 180 128 134 144 168 155 165  80 138
 142  90 197 172 116  85 163 178 171 103  83 140 194 146 192 106 135 153
  89

In [9]:
# STATISTICAL INFORMATION
pd.DataFrame(data = data.describe().T)

count        mean          std    min      25%     50%  \
battery_power  2000.0  1238.51850   439.418206  501.0   851.75  1226.0   
blue           2000.0     0.49500     0.500100    0.0     0.00     0.0   
clock_speed    2000.0     1.52225     0.816004    0.5     0.70     1.5   
dual_sim       2000.0     0.50950     0.500035    0.0     0.00     1.0   
fc             2000.0     4.30950     4.341444    0.0     1.00     3.0   
four_g         2000.0     0.52150     0.499662    0.0     0.00     1.0   
int_memory     2000.0    32.04650    18.145715    2.0    16.00    32.0   
m_dep          2000.0     0.50175     0.288416    0.1     0.20     0.5   
mobile_wt      2000.0   140.24900    35.399655   80.0   109.00   141.0   
n_cores        2000.0     4.52050     2.287837    1.0     3.00     4.0   
pc             2000.0     9.91650     6.064315    0.0     5.00    10.0   
px_height      2000.0   645.10800   443.780811    0.0   282.75   564.0   
px_width       2000.0  1251.51550   432.199447  500.0   874.75  1247.0   
ram            2000.0  2124.21300  1084.732044  256.0  1207.50  2146.5   
sc_h           2000.0    12.30650     4.213245    5.0     9.00    12.0   
sc_w           2000.0     5.76700     4.356398    0.0     2.00     5.0   
talk_time      2000.0    11.01100     5.463955    2.0     6.00    11.0   
three_g        2000.0     0.76150     0.426273    0.0     1.00     1.0   
touch_screen   2000.0     0.50300     0.500116    0.0     0.00     1.0   
wifi           2000.0     0.50700     0.500076    0.0     0.00     1.0   
price_range    2000.0     1.50000     1.118314    0.0     0.75     1.5   

                   75%     max  
battery_power  1615.25  1998.0  
blue              1.00     1.0  
clock_speed       2.20     3.0  
dual_sim          1.00     1.0  
fc                7.00    19.0  
four_g            1.00     1.0  
int_memory       48.00    64.0  
m_dep             0.80     1.0  
mobile_wt       170.00   200.0  
n_cores           7.00     8.0  
pc               15.00    20.0  
px_height       947.25  1960.0  
px_width       1633.00  1998.0  
ram            3064.50  3998.0  
sc_h             16.00    19.0  
sc_w              9.00    18.0  
talk_time        16.00    20.0  
three_g           1.00     1.0  
touch_screen      1.00     1.0  
wifi              1.00     1.0  
price_range       2.25     3.0

In [10]:
# Let us see the names of the columns
data.columns

Index(['battery_power', 'blue', 'clock_speed', 'dual_sim', 'fc', 'four_g',
       'int_memory', 'm_dep', 'mobile_wt', 'n_cores', 'pc', 'px_height',
       'px_width', 'ram', 'sc_h', 'sc_w', 'talk_time', 'three_g',
       'touch_screen', 'wifi', 'price_range'],
      dtype='object')

###We can see that sc_width column and px_height column has minimum value 0 which is not possible in any mobile. **We need to handle this mismatch**


---


# **HANDLING MISMATCH VALUES**

# In px_height column

In [11]:
# Checking observations having px_hieght value as 0.
print(f"number of observation having px_height as 0 = {data[data['px_height']==0].shape[0]}")
# Checking observations having screen width value as 0.
print(f"number of observation having screen width value as 0 = {data[data['sc_w']==0].shape[0]}")

number of observation having px_height as 0 = 2
number of observation having screen width value as 0 = 180


In [12]:
#observation of px_height  having value 0
data[data['px_height']==0]

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
1481           1834     0          2.1         0   7       1          40   
1933            897     1          2.0         0   3       1           2   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
1481    0.1         99        4  ...          0      1987  3692    13     0   
1933    0.6        154        8  ...          0       994  1958     7     5   

      talk_time  three_g  touch_screen  wifi  price_range  
1481         16        1             1     0            3  
1933          7        1             1     0            1  

[2 rows x 21 columns]

In [13]:
# As there are only 2 observations having px_height=0. so just drop it.
data.drop(data[data['px_height']==0].index, inplace=True)

In [14]:
# Rechecking the observation of px_height having value 0 after dropping 
data[data['px_height']==0]

Empty DataFrame
Columns: [battery_power, blue, clock_speed, dual_sim, fc, four_g, int_memory, m_dep, mobile_wt, n_cores, pc, px_height, px_width, ram, sc_h, sc_w, talk_time, three_g, touch_screen, wifi, price_range]
Index: []

[0 rows x 21 columns]

# In sc_width column

In [15]:
# Checking observations having screen width value as 0.
print(f"number of observation having screen width value as 0 = {data[data['sc_w']==0].shape[0]}")

number of observation having screen width value as 0 = 179


### Observation: Earlier it was 181,but as we drop rows in px_height having value 0 then 2 rows from it also dropped. That's why it is 179 now.

###We will use KNN imputer method for imputation

In [16]:
# Replacing 0 with NAN so that we can implement KNN Imputer.
data['sc_w']=data['sc_w'].replace(0,np.nan)

In [17]:
# import KNN imputer from sklearn
from sklearn.impute import KNNImputer
impute_knn = KNNImputer(n_neighbors=1)
data=pd.DataFrame(impute_knn.fit_transform(data),columns=data.columns)

In [18]:
# Checking shape again
data.shape

(1998, 21)

In [19]:
# Rechecking observations of 'screen width' whether the zero values are removed or not.
print(f"number of observation having screen width value as 0 = {data[data['sc_w']==0].shape[0]}")

number of observation having screen width value as 0 = 0


## Hence, we handled our mismatch column successfully